Upstream svgpathtools improvement - v2.1

NGExplorer opened the pull request:
Fixes https://github.com/mathandy/svgpathtools/issues/232

- Some GPU does not have 'arc' so we can use cubic bezier for such case.
- When GPU does not support 'arc' command Circle can be rendered using 2 arcs

Incorporated review comments mentioned in PR https://github.com/mathandy/svgpathtools/pull/233

@mathandy Please review and let us know if any changes are required.

mathandy commented:
Hi @NGExplorer, it looks like several of my requests / comments in #233 have not been taken care of yet. Please take a second look.

mathandy commented:
Also, would it be possible for you to add some test?

NGExplorer commented:
Hello @mathandy ,

On vglite-tools-2.1 branch we've incorporated following changes,
We rechecked your feedback on old PR #233 there were 4 review comments
1. Keep old behavior for 'arc' (keep use_cubics=False as default) to keep old toolkit behavior identical. ([link](https://github.com/mathandy/svgpathtools/pull/233/files/833a0a64c034cfe101196a47c6aab3683815bd9e))
    We have incorporated this change. ([fix](https://github.com/mathandy/svgpathtools/pull/236/commits/a48d21d5cc3965ab2cd890601126f2fb84f55a3a))

2. Use is_polygon variable in call to polyline2pathd
   We have incorporated this change. return polyline2pathd(polyline, is_polygon)
![image](https://github.com/user-attachments/assets/d4734938-8529-4770-b745-89c4c964d1d3)

3. points = [(0,0)] ([link](https://github.com/mathandy/svgpathtools/pull/233#pullrequestreview-2804627612))   
    Failing vector [shapes-polygon-02-t.svg](https://www.w3.org/Graphics/SVG/Test/20080912/svggen/shapes-polygon-02-t.svg)
![image](https://github.com/user-attachments/assets/5407bab3-5129-471f-8640-cb0989096f70)
    Expected output: https://www.w3.org/Graphics/SVG/Test/20080912/htmlEmbedHarness/shapes-polygon-02-t.html
    We have modified our application(gpu-vglite-toolkit) to not invoke polyline2pathd(), when there are no points.

5. Remove spaces around VG commands ([link](https://github.com/mathandy/svgpathtools/pull/233/files/1396f997f1716c93d995e6677a28e608971eae23))
    We have removed its original patch() which introduced this issue.

Let us know if we missed something to improve svgpathtools.

mathandy commented on svgpathtools/svg_to_paths.py:
```diff
@@ -27,31 +30,54 @@ def path2pathd(path):
 
-    return d + 'z'
+    
+    return d + ' Z'
```
Does this change have any effect?
(if no, let's remove it to avoid unnecessarily changing output)

NGExplorer replied:
No, this changes does not have side-effect. We will remove it.

mathandy commented on svgpathtools/svg_to_paths.py:
```diff
@@ -68,7 +94,7 @@ def polyline2pathd(polyline, is_polygon=False):
     # e.g. `parse_path('M0 0L100 100Z') == parse_path('M0 0L100 100L0 0Z')`
     # This check ensures that an n-point polygon is converted to an n-Line path.
-    if is_polygon and closed:
+    if is_polygon or closed:
```
What's the justification for this change?

I think this needs to be reverted, but if you have some reason for the change, we can talk about a solution

NGExplorer replied:
In SVG, a <polygon> element is always rendered as a closed shape, regardless of whether the last point explicitly matches the first point. For example:

Test vector: [shapes-polygon-01-t.svg](https://www.w3.org/Graphics/SVG/Test/20080912/svggen/shapes-polygon-01-t.svg)

`<polygon xml:id="polygon-01" fill="none" stroke="#000000" points="59,45,95,63,108,105,82,139,39,140,11,107,19,65" />
`
In this case, the first and last points differ, so a 'closed' flag computed from point equality would be false. 

The original condition:

if is_polygon and closed:
    points.append(points[0])

fails to append the starting point because closed flag is false, even though 'is_polygon' is True. This can lead to incorrect processing in cases where polygon closure needs to be explicitly handled in the logic.

To account for the SVG specification and ensure consistent behaviour, this condition should be updated to:

if is_polygon or closed:
    points.append(points[0])

This change ensures that <polygon> elements are always treated as closed paths, aligning with the SVG specification.

mathandy replied:
"This can lead to incorrect processing in cases where polygon closure needs to be explicitly handled in the logic."

@NGExplorer Can you give an example of this please?

NGExplorer replied:
> "This can lead to incorrect processing in cases where polygon closure needs to be explicitly handled in the logic."
> 
> @NGExplorer Can you give an example of this please?

I got your point,
rect() and polygon() both uses polyline2pathd()

closure is implicit in polygon and is_polygon already sets to True in such cases.
Where is in case of polyline is_polygon is false.

In vector drawing close command can be specified using 'z' or Line (origin) command.
We are trying to force 'z' command where possible.

As a result if we compute closed by taking is_polygon in consideration, original condition can be retained.
Patch may look something like following, We will check this in our test-suite and update you by EOD.
```
     closed = (float(points[0][0]) == float(points[-1][0]) and
          float(points[0][1]) == float(points[-1][1]))
+    closed = closed or (is_polygon == True)

if is_polygon and closed:
    points.append(points[0])
    

+   if closed:
-   if is_polygon or closed:
        d += 'z'

```

mathandy commented on svgpathtools/svg_to_paths.py:
```diff
@@ -77,22 +103,22 @@ def polyline2pathd(polyline, is_polygon=False):
     rectangle object and proceed counter-clockwise."""
-    x, y = float(rect.get('x', 0)), float(rect.get('y', 0))
-    w, h = float(rect.get('width', 0)), float(rect.get('height', 0))
+    x, y = float(rect.get('x', 0) or 0), float(rect.get('y', 0) or 0)
```
What's the justification for these defaults?

mathandy commented on svgpathtools/svg_to_paths.py:
```diff
@@ -27,31 +30,54 @@ def path2pathd(path):
 
-    cx = ellipse.get('cx', 0)
-    cy = ellipse.get('cy', 0)
+    cx = (ellipse.get('cx', 0) or 0)
```
You've made several changes like this -- what's the justification? Is null a valid option here? If so, why change it to zero?

NGExplorer replied:
Reverted this change.

mathandy commented:
> Hi @mathandy Could you please review the changes incorporated by Nilam? If there is something else to be done just let us know. Thank you!

Done.
Also, please add unit tests for any changes to logic (assuming you have justification for them and want to keep them), as well as the new feature added.

nicusorcitu commented on svgpathtools/svg_to_paths.py:
```diff
@@ -1,3 +1,6 @@
+#
+# Modified by NXP 2024
```
Hi @mathandy 
This is the modification copyright according to our NXP policy.

For the newly created files we have to add:
"Copyright 2025 NXP"

and just for those kind of files that already exist, but without copyright, we do not want to take any extra credit so this is why it says just:
"Modified by NXP 2025"

Please let me know if staying as is looks acceptable to you.

NGExplorer commented:
@mathandy Please review and let us know if any changes are required.

mathandy commented on svgpathtools/svg_to_paths.py:
```diff
@@ -1,3 +1,6 @@
+#
+# Modified by NXP 2024, 2025
```
Did you leave this attribution in by accident or is there some reason you really want it?

nicusorcitu replied:
hi @mathandy 
I left an explanation here: https://github.com/mathandy/svgpathtools/pull/236#discussion_r2092810815

Is this acceptable to stay in the code? It just says that we, NXP, modified some code on the offending file. It is NXP policy to enforce this upstream.

nicusorcitu replied:
OK, we moved our modified copyright on the if block.
Thanks!

mathandy commented:
Looks great! I asked a question about the attribution, other than that, I think this is ready.

mathandy commented:
Thanks @nicusorcitu and @NGExplorer !
